### PR TITLE
Update of Protein Parsimony to accommodate multiple Proteases

### PR DESCRIPTION
**Pull request by rmmiller22**
Allows for Protein Parsimony to account for which protease was used to produce this peptide when performing parsimony.  This code has been tested.

Also, edit was made to digestion params constructor so that the Protease object can be used instead of a string representation of protease name. This lead to small changes in many files changing from protease.name to protease. Also eliminated all of the unnecessary protease additions to the global protease dictionary with this fix.

### CODE REVIEW

File: EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs [hunk 1]
```diff
@@ -48,7 +48,15 @@ protected static void ResolveAmbiguities(Dictionary<CompactPeptideBase, HashSet<
             foreach (CompactPeptide key in keys)
             {
                 HashSet<PeptideWithSetModifications> value = compactPeptideToProteinPeptideMatching[key];
-                compactPeptideToProteinPeptideMatching[key] = new HashSet<PeptideWithSetModifications> { value.FirstOrDefault(b => !b.Protein.IsDecoy) ?? value.First() };
+                HashSet<PeptideWithSetModifications> resolvedValues = new HashSet<PeptideWithSetModifications>();
```
Comment (rmillikin) [inline EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs]: revert

File: EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs [hunk 2]
```diff
@@ -74,20 +82,18 @@ protected override MetaMorpheusEngineResults RunSpecific()
             double proteinsMatched = 0;
             int oldPercentProgress = 0;
 
-            Parallel.ForEach(Partitioner.Create(0, proteins.Count), fff =>
-            {
-                for (int i = fff.Item1; i < fff.Item2; i++)
+
```
Comment (rmillikin) [inline EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs]: revert to parallelized, we should investigate if there's a bug in the parallelization code (it's super slow in some cases without parallelization)

File: EngineLayer/DigestionParams.cs [hunk 3]
```diff
@@ -18,15 +18,23 @@ public class DigestionParams
...
         // this parameterless constructor needs to exist to read the toml. 
         // if you can figure out a way to get rid of it, feel free...
-        public DigestionParams() : this("trypsin")
+        public DigestionParams() : this(GlobalVariables.ProteaseDictionary["trypsin"])
         {
 
         }
 
-        public DigestionParams(string protease = "trypsin", int MaxMissedCleavages = 2, int MinPeptideLength = 7, int MaxPeptideLength=int.MaxValue, int MaxModificationIsoforms = 1024, 
+        public DigestionParams(Protease protease = null, int MaxMissedCleavages = 2, int MinPeptideLength = 7, int MaxPeptideLength=int.MaxValue, int MaxModificationIsoforms = 1024,
```
Comment (rmillikin) [inline EngineLayer/DigestionParams.cs]: revert

File: EngineLayer/PeptideSpectralMatch.cs [hunk 4]
```diff
@@ -15,6 +15,7 @@ public class PeptideSpectralMatch
         #region Private Fields
 
         private const double tolForDoubleResolution = 1e-6;
+        private DigestionParams DigestionParam { get; set; }
```
Comment (rmillikin) [inline EngineLayer/PeptideSpectralMatch.cs]: delete, rename line 92 to what you want

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 5]
```diff
@@ -14,7 +14,7 @@ public class ProteinParsimonyEngine : MetaMorpheusEngine
 
         private readonly bool treatModPeptidesAsDifferentPeptides;
         private readonly Dictionary<CompactPeptideBase, HashSet<PeptideWithSetModifications>> compactPeptideToProteinPeptideMatching;
-
+        //dictionary with key being object contianing mass of the compact peptide and value being a hasset of peptides with set modifications having that mass
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: comment is fine, put it before the thing you're commenting

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 6]
```diff
@@ -45,11 +45,9 @@ protected override MetaMorpheusEngineResults RunSpecific()
 
         private List<ProteinGroup> ApplyProteinParsimony()
         {
-
             if (!compactPeptideToProteinPeptideMatching.Values.Any())//if dictionary is empty return an empty list of protein groups
-            {
                 return new List<ProteinGroup>();
-            }
+
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: put brackets back

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 7]
```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
 
             foreach (var kvp in compactPeptideToProteinPeptideMatching)
             {
-                // finds unique peptides (peptides that can belong to only one protein)
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: revert

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 8]
```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
...
-                    }
-                    else
-                    {
-                        peptides.UnionWith(kvp.Value);
-                    }
-                }
-
                 // if a peptide is associated with a decoy protein, remove all target protein associations with the peptide
                 if (kvp.Value.Any(p => p.Protein.IsDecoy))
-                {
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: keep brackets

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 9]
```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
...
+               if (proteinsAssociatedWithThisPeptide.Count == 1)
+               {
+                  if (!proteinsWithUniquePeptides.TryGetValue(kvp.Value.First().Protein, out HashSet<PeptideWithSetModifications> peptides))
+                      proteinsWithUniquePeptides.Add(kvp.Value.First().Protein, new HashSet<PeptideWithSetModifications>(kvp.Value));
+                  else
+                      peptides.UnionWith(kvp.Value);
+               }
+               else
+               {
+                    PeptideWithSetModifications[] peptides = new PeptideWithSetModifications[kvp.Value.Count];
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: var peptides = kvp.Value.ToArray();

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 10]
```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
...
+                    int i = 0;
+                    foreach (var peptide in kvp.Value)
+                    {
+                        peptides[i] = peptide;
+                        i++;
+                    }
+
+                    foreach (var peptide in kvp.Value)
+                    {
+                        Protease protease = peptide.digestionParams.Protease;
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: // for this peptide with set mods, get its protease
// if there are other peptides with the same protease, it's shared
// if not, it's unique for that protease (and therefore counted as unique for parsimony)

File: EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs [hunk 11]
```diff
@@ -120,33 +117,62 @@ private List<ProteinGroup> ApplyProteinParsimony()
...
+                        int j = 0;
```
Comment (rmillikin) [inline EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs]: int sameProteaseCount = kvp.Value.Count(v => v.digestionParams.Protease == peptide.digestionParams.protease);

File: EngineLayer/Proteomics/Peptide.cs [hunk 12]
```diff
@@ -191,7 +191,7 @@ public string BaseSequence
                         kvp.Add(ok.Key, ok.Value);
                     }
                 }
-                yield return new PeptideWithSetModifications(Protein, OneBasedStartResidueInProtein, OneBasedEndResidueInProtein,
+                yield return new PeptideWithSetModifications(Protein, digestionParams, OneBasedStartResidueInProtein, OneBasedEndResidueInProtein,
```
Comment (rmillikin) [inline EngineLayer/Proteomics/Peptide.cs]: check with Anthony that this change carries over to mzLib if relevant

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 13]
```diff
@@ -59,6 +61,14 @@ public PeptideWithSetModifications(PeptideWithSetModifications modsFromThisOne,
             this.numFixedMods = numFixedMods;
             this.allModsOneIsNterminus = allModsOneIsNterminus ?? new Dictionary<int, ModificationWithMass>();
         }
+        public PeptideWithSetModifications(Protein protein, DigestionParams digestionParams, int oneBasedStartResidueInProtein, int oneBasedEndResidueInProtein, string peptideDescription, int missedCleavages,
```
Comment (rmillikin) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: this is fine, but try removing the other constructor and see what breaks and talk to me about it

File: TaskLayer/MetaMorpheusTask.cs [hunk 14]
```diff
@@ -345,7 +345,26 @@ protected static void WritePsmsToTsv(IEnumerable<PeptideSpectralMatch> items, st
                 {
                     output.WriteLine(heh.ToString(ModstoWritePruned));
                 }
-            }
+           }
+        }
+        protected static HashSet<DigestionParams> GetListOfDistinctDigestionParams(CommonParameters commonParameters, IEnumerable<CommonParameters> enumerable)
```
Comment (rmillikin) [inline TaskLayer/MetaMorpheusTask.cs]: check to see if you use this; if not, remove it

File: TaskLayer/SearchTask/SearchTask.cs [hunk 15]
```diff
@@ -37,6 +37,7 @@ public SearchTask() : base(MyTask.Search)
 
         #region Public Methods
 
+
```
Comment (rmillikin) [inline TaskLayer/SearchTask/SearchTask.cs]: remove extra spaces

File: Test/MultiProteaseParsimonyTest.cs [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+            ProteinParsimonyEngine ppe = new ProteinParsimonyEngine(CompactPeptidesToProteinPeptidesMatching, false, null);
+            var proteinAnalysisResults = (ProteinParsimonyResults)ppe.Run();
+          
+            List < ProteinGroup > proteinGroups = proteinAnalysisResults.ProteinGroups;
+            Assert.AreEqual(2, proteinGroups.Count);
+            Assert.AreEqual(2, proteinGroups.ElementAt(0).AllPeptides.Count);
+            Assert.AreEqual(1, proteinGroups.ElementAt(0).UniquePeptides.Count);
+            Assert.AreEqual(4, proteinGroups.ElementAt(1).AllPeptides.Count);
+            Assert.AreEqual(3, proteinGroups.ElementAt(1).UniquePeptides.Count);
+
```
Comment (rmillikin) [inline Test/MultiProteaseParsimonyTest.cs]: assert that the peptide sequences in AllPeptides and UniquePeptides are what you expect

File: EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs (hunk 1 above)
Comment (rmmiller22) [reply EngineLayer/CompactToProteinMatchEngine/SequencesToActualProteinPeptidesEngine.cs]: If i revert this then parsimony will fail

**Pull request merged**